source/cl/source/extension/source/intel_unified_shared_memory/intel_unified_shared_memory.cpp

```diff
@@ -418,6 +424,141 @@ cl_int device_allocation_info::allocate(cl_uint alignment) {
 
   return CL_SUCCESS;
 }
+
+shared_allocation_info::shared_allocation_info(const cl_context context,
+                                               const cl_device_id device,
+                                               const size_t size)
+    : allocation_info(context, size),
+      device(device),
+      mux_memory(nullptr),
+      mux_buffer(nullptr) {
+  if (device) {
+    cl::retainInternal(device);
+  }
+};
+
+shared_allocation_info::~shared_allocation_info() {
+  if (mux_buffer) {
+    assert(device);
+    (void)muxDestroyBuffer(device->mux_device, mux_buffer,
+                           device->mux_allocator);
+  }
+
+  if (mux_memory) {
+    assert(device);
+    muxFreeMemory(device->mux_device, mux_memory, device->mux_allocator);
+  }
+
+  // Free the host side allocation
+  if (base_ptr) {
+    cargo::free(base_ptr);
+  }
+
+  if (device) {
+    cl::releaseInternal(device);
+  }
+};
+
+cargo::expected<std::unique_ptr<shared_allocation_info>, cl_int>
+shared_allocation_info::create(cl_context context, cl_device_id device,
+                               const cl_mem_properties_intel *properties,
+                               size_t size, cl_uint alignment) {
+  OCL_CHECK(size == 0, return cargo::make_unexpected(CL_INVALID_BUFFER_SIZE));
+
+  OCL_CHECK((alignment & (alignment - 1)) != 0,
+            return cargo::make_unexpected(CL_INVALID_VALUE));
+
+  if (device) {
+    const auto device_align = device->mux_device->info->buffer_alignment;
+    OCL_CHECK(size > device->max_mem_alloc_size,
+              return cargo::make_unexpected(CL_INVALID_BUFFER_SIZE));
+    OCL_CHECK(alignment > device_align,
+              return cargo::make_unexpected(CL_INVALID_VALUE));
+
+    if (alignment == 0) {
+      alignment = device_align;
+    }
+  }
+
+  auto alloc_properties = parseProperties(properties, true);
+  if (!alloc_properties) {
+    return cargo::make_unexpected(alloc_properties.error());
+  }
+
+  // Default alignment with no device is the highest alignment of all other
+  // devices
+  if (!device && alignment == 0) {
+    cl_uint max_align = 1;
+    for (auto device : context->devices) {
+      const auto device_align = device->min_data_type_align_size;
+      max_align = device_align > max_align ? device_align : max_align;
+    }
+    alignment = max_align;
+  }
+
+  auto usm_alloc = std::unique_ptr<shared_allocation_info>(
+      new (std::nothrow) shared_allocation_info(context, device, size));
```

Comment:
This uses `new (std::nothrow)` allowing a `nullptr` return, but then unconditionally dereferences the result.